File: crack-data-structures-and-algorithms/leetcode/python-impl/minimum_path_sum_q64.py
class Solution(object):
    def minPathSum(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        if len(grid) == 0:
            return 0

        rcnt = len(grid)
        ccnt = len(grid[0])
        m = [[0] * ccnt for _ in range(rcnt)]
        m[0][0] = grid[0][0]

        for i in range(1, ccnt):
            m[0][i] = m[0][i-1] + grid[0][i]

        for i in range(1, rcnt):
            m[i][0] = m[i-1][0] + grid[i][0]

        for i in range(1, rcnt):
            for j in range(1, ccnt):
                m[i][j] = min(m[i-1][j], m[i][j-1]) + grid[i][j]

        return m[-1][-1]

File: crack-data-structures-and-algorithms/leetcode/minimum_path_sum_q64.cpp
#include <algorithm>
#include <vector>

using namespace std;

// 核心思路
// 令 dp[i][j] 代表走到 (i,j) 的最小路径和，则我们有
// dp[0][0] = grid[0][0]
// dp[i][0] = dp[i-1][0] + grid[i][0], 1 <= i < rows
// dp[0][j] = dp[0][j-1] + grid[0][j], 1 <= j < cols
// dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j], 1 <= i, j < rows/cols
// 然后根据状态方程求解即可
class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        int rows = grid.size();
        int cols = grid[0].size();

        vector<vector<int>> dp(rows, vector<int>(cols));
        dp[0][0] = grid[0][0];

        for (auto i = 1; i < rows; ++i) {
            dp[i][0] = dp[i-1][0] + grid[i][0];
        }

        for (auto i = 1; i < cols; ++i) {
            dp[0][i] = dp[0][i-1] + grid[0][i];
        }

        for (auto i = 1; i < rows; ++i) {
            for (auto j = 1; j < cols; ++j) {
                dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j];
            }
        }

        return dp.back().back();
    }
};
